actioncable/lib/action_cable/connection/web_socket_rack.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module ActionCable
+  module Connection
+    # Wrap the real socket to minimize the externally-presented API
+    #
+    # This variation leverages the `rack.upgrade?` approach detailed here: https://github.com/rack/rack/pull/1272
+    class WebSocketRack # :nodoc:
+      UPGRADE_EXISTS = "rack.upgrade?"
+      UPGRADE = "rack.upgrade"
+      PROTOCOL_NAME_IN = "HTTP_SEC_WEBSOCKET_PROTOCOL"
+      PROTOCOL_NAME_OUT = "Sec-Websocket-Protocol"
+      CLOSE_REASON = ""
+
+      attr_reader :protocol
+
+      def initialize(env, event_target, event_loop, protocols)
+        env[UPGRADE] = self
+        @event_target = event_target
+        @protocol = nil
+        @websocket = nil
+        if env[PROTOCOL_NAME_IN]
```

Comment:
Can we remove the outer `if` and just leave `if env[...].is_a?(String) ... elseif env[...].is_a?(Array)`?

Or add a guard close to avoid nested `if`-s: `return unless env[...]`

Reply:
Or, probably, re-write it the following way to avoid iteration duplication:

```ruby
request_protocols = env[PROTOCOL_NAME_IN]
return if request_protocols.nil?

request_protocols = request_protocols.split(/,\s?/) if request_protocols.is_a?(String)

request_protocols.each do |request_protocol|
  break @protocol = request_protocol if protocols.include?(request_protocol)
end
```

Reply:
This is a good point.

I'll adopt your suggestion.